src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -302,7 +390,72 @@ TransactionManager::rewrite_extent_ret TransactionManager::rewrite_extent(
     cache->duplicate_for_write(t, extent);
     return rewrite_extent_iertr::now();
   }
-  return lba_manager->rewrite_extent(t, extent);
+
+  return lba_manager->rewrite_physical_extent(t, extent);
+}
+
+TransactionManager::rewrite_extent_ret TransactionManager::rewrite_extents(
```

Comment:
I'm still reading this PR in detail, but I specifically do not want to delay performing OOL writes until submit_transaction.  Wasted writes in the event of a retry shouldn't be a concern as we need to minimize retries anyway.  Permitting OOL writes prior to submit_transaction should permit far more device level parallelism in the event that we don't conflict (which should be the common case).

Reply:
> but I specifically do not want to delay performing OOL writes until submit_transaction.

I was not talking about when to perform ool writes. I modified my comment above, maybe it is clearer now.

Reply:
Ah, sorry, was referring to a conversation I'd had with @xuxuehan .  Yeah, I'm working on some code to implement that now.  At first, I'd wanted to do something simpler for this PR, but this PR ends up having to replace paddr mappings anyway, so we might as well do it generally for all users.

Reply:
Yeah, for all users.

Reply:
And this is why I think EPM is responsible to decide whether to inline or not according to hint @xxhdx1985126 .

Reply:
I see, so we've decided that EPM should be the decision maker for the inline-ness of extents, right? @cyx1231st @athanatos 

> Yeah, I'm working on some code to implement that now. At first, I'd wanted to do something simpler for this PR, but this PR ends up having to replace paddr mappings anyway, so we might as well do it generally for all users.

If so, I think I can do this right now, so that the whole functionality can be in position more quickly.

Reply:
> I think we were talking about applying ool-extents to the general ~submit_transaction()~ transactions according to [#41803 (comment)](https://github.com/ceph/ceph/pull/41803#discussion_r655713157) cc @athanatos, not only about the cleaner transactions in `rewrite_extents()` only.

Sorry, I didn't get your point at first, I totally agree with you, as I stated in [#42270 ](https://github.com/ceph/ceph/pull/42270#issuecomment-881486195).

Reply:
Physical extents should be be allocated directly as inline extents for now -- no need to modify anything below TransactionManager.  TransactionManager should use a new Cache::alloc_new_extent_delayed call to create logical extents (whether during rewrite or in TransactionManager::alloc_extent) in a Transaction::delayed_alloc list to be considered/written by EPM in TransactionManager::submit_transaction _direct prior to entering the prepare phase (and thus prior to Cache::prepare_record).

EPM probably shouldn't be involved prior to TransactionManager::submit_transaction_direct -- delaying the choice of where to allocate the extents would allow us to be smarter about placement later.

Reply:
The delayed alloc extents should have a magic P_ADDR_DELAYED value for the physical address in the mean time between the initial Cache::alloc_new_extent_delayed call and Transaction::submit_transaction_direct, I think.  I don't see a reason to increment the offset between delayed allocations since they don't necessarily end up in the same segment and we won't ultimately write these down as relative addrs.

Reply:
After EPM for each extent either writes out or adds to the inline record set, you'll need to call into the lba manager to update the lba->physical mapping to reflect either the new record/extent-relative addr for inline or the absolute addr for ool (LBAManager::update_logical_extent_mapping).

Reply:
Later on, we'll want to add some additional in-memory structures to avoid the second lookup to fix the lba leaves, but this will be fine for now.

Reply:
@athanatos Just to make sure I'm getting this right, I think what you are suggesting is as follows:

1. inline-ness of logical extents are decided by EPM;
2. we delay the address allocation of logical extents (whether inline or ool) until we enter `submit_transaction_direct` but before entering `prepare_record`, so that we still  have the opportunity to update lba mapping;
3. the reason for delay the involvement of EPM is that the later EPM is involved, the more information about extents EPM can get, and so the smarter it can be.

Am I right?

Reply:
1. Yep
2. Right, and also to do the writes prior to the prepare pipeline phase which is exclusive.
3. Right, as with any existing inline extent.
4. Yeah.  For example, with this implementation we need to write a record header for each group of ool extents requiring at least one block.  A transaction with two newly allocated 4k extents should really never split them across two ool segments simply due to the extra metadata overhead regardless of the hints on the extents.  No need to actually worry about that for now, but delaying selection of an allocator/segment until submit_transaction_direct would give us enough information to address this later on (along with other things) and doesn't seem to be me to be any more complicated.  This does mean you'll need to stash the allocation hint in the delayed_alloc list with each extent.

Reply:
Got it, consider it done:-)

Reply:
> delaying the choice of where to allocate the extents would allow us to be smarter about placement later.

Hmm, I think delaying the choice doesn't work for PMem, the decision needs to be made whether to allocate memory  from RAM or PMem for `CachedExtent::ptr` according to the user hint prior to the prepare phase.

And I cannot think of an advantages for RBM too, as the paddr can be immediately available from its block allocator if it should be an ool-extent according to the hint.

So I don't really see why we need to delay the involvement of EPM until transaction submission. Seems to me the only reason is we need to delay the paddr assignments for inline-extents and segmented ool-extents, but that logic can be independent from EPM decesions.

Does it make sense?

Reply:
> Hmm, I think delaying the choice doesn't work for PMem, the decision needs to be made whether to allocate memory from RAM or PMem for `CachedExtent::ptr` according to the user hint prior to the prepare phase.

I think you mean that, for PMem, we can't delay the address allocation because the mem address is the eventual storage address，right？

Reply:
I don't think it matters either way for RBM.  Pmem is, however, different as you point out, but I'd rather deal with that in a future PR.

Reply:
> I don't think it matters either way for RBM. Pmem is, however, different as you point out, but I'd rather deal with that in a future PR.

I think maybe we can deal with it this wan:

1. first, EPM provide the alloc_new_extent interface, in which it decides whether to call `Cache::alloc_new_extent` or `Cache::alloc_new_extent_delayed`, based on the decision whether the extent should go to PMem or somewhere else;
2. all extent allocations go through EPM, as EPM gets to decide where to allocation the memory space, from RAN or PMem;
3. if an extent is to go to PMem, then no other allocation operation is needed, otherwise, we do as mentioned [above](https://github.com/ceph/ceph/pull/41803#discussion_r674500971)

Does this make sense?:-)

Reply:
Sure.

Reply:
> I don't think it matters either way for RBM. Pmem is, however, different as you point out, but I'd rather deal with that in a future PR.

I'm OK, we can deal with it until necessary.

Reply:
I'm ok with adding a stub as suggested by @xxhdx1985126, but I suspect that Pmem will live mainly inside of Cache.  Pmem won't require an IO as such for allocation or writes, so I'm not entirely sure it'll need EPM, or if it does the pathway may not look much like that stub.  Either way is ok by me as long as it's easy to read.

Reply:
> I'm ok with adding a stub as suggested by @xxhdx1985126, but I suspect that Pmem will live mainly inside of Cache. Pmem won't require an IO as such for allocation or writes, so I'm not entirely sure it'll need EPM, or if it does the pathway may not look much like that stub. Either way is ok by me as long as it's easy to read.

Um, I don't quite follow, I thought EPM needs to decide which backend device the extent should go to when there are multiple backend devices. And before that we don't know whether an extent should be in PMem or not at all.

I mean, after the allocation, PMem extents don't need EPM any more; but before that we need EPM to choose whether those extents are PMem extents.  Am I right?

Reply:
Yeah, if we want EPM to have the logic for determining that the allocation would have to go through EPM.  Sounds good to me.